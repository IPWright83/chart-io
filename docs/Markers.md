# `<Markers />` Component API

Used to represent markers on a chart, these are useful for indicating the current mouse location.

## Structure

The markers that are rendered are pulled from the chart's Redux store. The interface of a marker is as follows:
| Field | Type | Note|
| ------- | ------------- | ------------------------------------------------------------------------------------------------------------------------------------------ |
| `cx` | `number` | This is the x-coordinate of the marker |
| `cy` | `number` | This is the y-coordinate of the marker |
| `fill` | `string` | This is the color of the marker |
| `r1` | `number` | This is the start radius of the marker. The marker will animate in size to the r2 value. |
| `r2` | `number` | This is the end radius of the marker |

## Props

| Prop    | Type          | Default | Note                                                                                                                                       |
| ------- | ------------- | ------- | ------------------------------------------------------------------------------------------------------------------------------------------ |
| `layer` | `HTMLElement` |         | This is an internal field and should not need setting manually. It is populated by the `withSVG` and `withCanvas` higher order components. |
