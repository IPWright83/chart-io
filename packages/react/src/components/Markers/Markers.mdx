import { Story, Canvas } from "@storybook/addon-docs/blocks";

# Markers Component

The `<Markers>` component allows us to define a layer to render markers on top of the plots. This is useful in highlighting a particular data point and can be combined with the `<Droplines>` layer for additional affect.

<Canvas>
    <Story id="components-markers--default" />
</Canvas>

### Props

| Prop      | Type          | Default | Note                                                                                                         |
| --------- | ------------- | ------- | ------------------------------------------------------------------------------------------------------------ |
| `layer`\* | `HTMLElement` |         | This is an _internal_ field and should not need setting manually. It is populated `withSVG` or `withCanvas`. |

<br />

## Adding Markers Programatically

To add a marker programatically you should call the `eventActions.addMarker` function and `eventActions.removeMarker` function for clean up:

```
const marker = {
    fill: "steelblue",
    stroke: "lightblue",
    cx: 50,
    cy: 50,
    r1: 10,
    r2: 30,
};

eventActions.addMarker(marker);
eventActions.removeMarker(marker);  // clean up
```

### Markers Schema:

| Prop       | Type     | Note                                                              |
| ---------- | -------- | ----------------------------------------------------------------- |
| `fill`     | `string` | The fill color of the marker                                      |
| `stroke`   | `string` | The stroke color of the marker                                    |
| **`cx`\*** | `number` | The x-coordinate of the marker                                    |
| **`cy`\*** | `number` | The y-coordinate of the marker                                    |
| `r1`       | `number` | The start radius of the marker (used to animate a growing marker) |
| **`r2`\*** | `number` | The end radius of the marker                                      |
