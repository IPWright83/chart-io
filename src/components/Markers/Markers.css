.chart-it.marker {
    stroke-width: 2;
    pointer-events: none;
}
