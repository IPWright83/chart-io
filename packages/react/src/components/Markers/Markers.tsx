import "./Markers.css";

import * as d3 from "d3";
import PropTypes from "prop-types";
import { useEffect } from "react";
import { useSelector } from "react-redux";

import { chartSelectors, eventSelectors, IStore } from "../../store";

export interface IMarkersProps {
    /**
     * The layer to be rendered upon. Typically this is an `<svg:g>` or a fake HTMLElement when using canvas.
     * @default undefined
     */
    layer: React.MutableRefObject<JSX.Element>;
}

/**
 * This component renders the markers that are triggered from various plots
 * @return {Component} The Markers component
 */
const Markers = ({ layer }: IMarkersProps) => {
    const animationDuration = useSelector((s: IStore) => chartSelectors.animationDuration(s));
    const theme = useSelector((s: IStore) => chartSelectors.theme(s));
    const markers = useSelector((s: IStore) => eventSelectors.markers(s));

    useEffect(() => {
        if (!layer.current) return;

        const join = d3.select(layer.current).selectAll(".marker").data(markers);

        join.exit().remove();

        // Add any new markers
        const enter = join
            .enter()
            .append("circle")
            .attr("class", "chart-it marker")
            .style("stroke", (d) => d.stroke || theme.markers.stroke)
            .style("stroke-width", theme.markers.strokeWidth)
            .style("filter", (d) => (theme.markers.shadow ? `drop-shadow(0px 0px 10px ${d.fill})` : undefined))
            .style("fill", (d) => d.fill || "none")
            .style("point-events", "none")
            .attr("r", (d) => d.r1 ?? d.r2 ?? theme.markers.size)
            .attr("cx", (d) => d.cx)
            .attr("cy", (d) => d.cy);

        enter
            .merge(join)
            .attr("cx", (d) => d.cx)
            .attr("cy", (d) => d.cy)
            .transition()
            .duration(animationDuration)
            .attr("r", (d) => d.r2 ?? theme.markers.size);
    }, [animationDuration, layer, markers]);

    return null;
};

export { Markers };
