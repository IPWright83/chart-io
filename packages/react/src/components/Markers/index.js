import { withSVG } from "../../hoc";
import { Markers as MarkersBase } from "./Markers";

const Markers = withSVG(MarkersBase, "markers");

export { Markers };
