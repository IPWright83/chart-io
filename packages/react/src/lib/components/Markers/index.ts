import { withSVG } from "../../hoc";
import { Markers as MarkersBase, IMarkersBaseProps } from "./Markers";

export interface IMarkersProps extends Omit<IMarkersBaseProps, "layer"> {}

export const Markers = withSVG<IMarkersProps>(MarkersBase, "markers");
