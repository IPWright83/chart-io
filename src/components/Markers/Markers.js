import "./Markers.css";

import * as d3 from "d3";
import PropTypes from "prop-types";
import { useEffect } from "react";
import { useSelector } from "react-redux";

import { chartSelectors, eventSelectors } from "../../store";

/**
 * This component renders the markers that are triggered from various plots
 * @return {Component} The Markers component
 */
const Markers = ({ layer }) => {
    const animationDuration = useSelector((s) => chartSelectors.animationDuration(s));
    const markers = useSelector((s) => eventSelectors.markers(s));

    useEffect(() => {
        if (!layer.current) return;

        const join = d3.select(layer.current).selectAll(".marker").data(markers);

        join.exit().remove();

        // Add any new markers
        const enter = join
            .enter()
            .append("circle")
            .attr("class", "marker")
            .style("stroke", (d) => d.stroke || "none")
            .style("fill", (d) => d.fill || "none")
            .attr("r", (d) => d.r1)
            .attr("cx", (d) => d.cx)
            .attr("cy", (d) => d.cy);

        enter
            .merge(join)
            .attr("cx", (d) => d.cx)
            .attr("cy", (d) => d.cy)
            .transition()
            .duration(animationDuration)
            .attr("r", (d) => d.r2);
    }, [animationDuration, layer, markers]);

    return null;
};

Markers.propTypes = {
    /**
     * The layer to be rendered upon. Typically this is an `<svg:g>` or a fake HTMLElement when using canvas.
     * @default undefined
     * @type {HTMLElement}
     */
    layer: PropTypes.object,
};

export { Markers };
