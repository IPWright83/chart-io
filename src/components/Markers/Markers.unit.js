import React from "react";
import { Provider } from "react-redux";
import TestRenderer from "react-test-renderer";

import { Markers } from "./Markers";

describe("Markers", () => {
    const store = {
        getState: () => ({
            event: {
                markers: [{ fill: "red", stroke: "blue", r1: 5, r2: 5, cx: 50, cy: 50 }],
            },
        }),
        dispatch: () => {},
        subscribe: () => {},
    };

    it("should render correctly", () => {
        const layer = { current: document.createElement("custom") };

        TestRenderer.act(() => {
            TestRenderer.create(
                <Provider store={store}>
                    <Markers layer={layer} />
                </Provider>
            ).toJSON();
        });

        expect(layer.current).toMatchSnapshot();
    });

    it("should skip if there is no layer avaliable", () => {
        const layer = { current: null };
        TestRenderer.act(() => {
            TestRenderer.create(
                <Provider store={store}>
                    <Markers layer={layer} />
                </Provider>
            ).toJSON();
        });

        // Should be empty
        expect(layer.current).toMatchSnapshot();
    });
});
