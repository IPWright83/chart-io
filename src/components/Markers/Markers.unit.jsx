import React from "react";
import { Provider } from "react-redux";

import { render } from "@testing-library/react";

import { Markers } from "./Markers";

describe("Markers", () => {
    const store = {
        getState: () => ({
            event: {
                markers: [{ fill: "red", stroke: "blue", r1: 5, r2: 5, cx: 50, cy: 50 }],
            },
        }),
        dispatch: () => {},
        subscribe: () => {},
    };

    it("should render correctly", () => {
        const layer = { current: document.createElement("custom") };

        render(
            <Provider store={store}>
                <Markers layer={layer} />
            </Provider>,
        );

        expect(layer).toMatchSnapshot();
    });

    it("should skip if there is no layer avaliable", () => {
        const layer = { current: null };

        render(
            <Provider store={store}>
                <Markers layer={layer} />
            </Provider>,
        );

        // Should be empty
        expect(layer).toMatchSnapshot();
    });
});
