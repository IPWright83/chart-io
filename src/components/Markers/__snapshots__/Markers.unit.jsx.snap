// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markers should render correctly 1`] = `
Object {
  "current": <custom>
    <circle
      class="marker"
      cx="50"
      cy="50"
      r="5"
      style="stroke: blue; fill: red;"
    />
  </custom>,
}
`;

exports[`Markers should skip if there is no layer avaliable 1`] = `
Object {
  "current": null,
}
`;
